RFC-0010-jdk-upgrade.md

```diff
@@ -0,0 +1,305 @@
+   |  \- net.jodah:failsafe:jar:2.0.1:runtime
+   +- com.fasterxml.jackson.core:jackson-annotations:jar:2.15.4:runtime
+   +- com.google.guava:guava:jar:32.1.0-jre:runtime
+   |  +- com.google.guava:failureaccess:jar:1.0.1:runtime
+   |  +- com.google.guava:listenablefuture:jar:9999.0-empty-to-avoid-conflict-with-guava:runtime
+   |  +- org.checkerframework:checker-qual:jar:3.37.0:runtime
+   |  +- com.google.errorprone:error_prone_annotations:jar:2.18.0:runtime
+   |  \- com.google.j2objc:j2objc-annotations:jar:2.8:runtime
+   +- com.google.inject:guice:jar:4.2.2:runtime
+   |  \- aopalliance:aopalliance:jar:1.0:runtime
+   +- javax.annotation:javax.annotation-api:jar:1.3.2:runtime
+   +- javax.inject:javax.inject:jar:1:runtime
+   +- com.facebook.airlift:concurrent:jar:0.216:runtime
+   +- com.facebook.airlift:configuration:jar:0.216:runtime
+   |  +- org.apache.bval:bval-jsr:jar:2.0.0:runtime
+   |  \- cglib:cglib-nodep:jar:3.2.5:runtime
+   +- com.facebook.airlift:bootstrap:jar:0.216:runtime
+   +- com.facebook.airlift:node:jar:0.216:runtime
+   +- com.facebook.airlift:log:jar:0.216:runtime
+   +- com.facebook.airlift:log-manager:jar:0.216:runtime
+   |  +- org.slf4j:slf4j-jdk14:jar:1.7.32:runtime
+   |  +- org.slf4j:log4j-over-slf4j:jar:1.7.32:runtime
+   |  \- org.slf4j:jcl-over-slf4j:jar:1.7.32:runtime
+   +- com.google.code.findbugs:jsr305:jar:3.0.2:runtime
+   +- org.pcollections:pcollections:jar:2.1.2:runtime
+   +- org.weakref:jmxutils:jar:1.19:runtime
+   +- org.openjdk.jol:jol-core:jar:0.2:runtime
+   +- joda-time:joda-time:jar:2.12.7:runtime
+   +- javax.validation:validation-api:jar:2.0.1.Final:runtime
+   +- it.unimi.dsi:fastutil:jar:8.5.2:runtime
+   \- org.apache.commons:commons-text:jar:1.10.0:runtime
+      \- org.apache.commons:commons-lang3:jar:3.12.0:runtime
+
+```  
+  
+</details>
+
+By carefully limiting how and when a dependency is upgraded, we can ensure that Spark jobs using Presto are still runnable on older Spark clusters
+
+For example, the Jetty dependency comes in via `presto-main`. If we can figure out a way to either  
+
+1. Remove this dependency or   
+1. Replace it with an equivalent implementation,  
+
+we can migrate `presto-main` to a higher JDK version
+In this particular case, we could identify that code in PoS that uses Jetty's HttpClient can be deprecated. See [TODO issue link]() for more details on this
+
+For the modules that *cannot be upgraded* to a higher bytecode level, we can refactor code to generate smaller modules that can be kept at lower bytecode levels.  
+This can prove to be beneficial for other modules too, e.g `presto-client` can be kept at JVM 8 bytecode level to allow for a broader support for Java based clients
+
+## Timeline for the project
+
+A detailed timeline is out-of-scope for this RFC.  
+We will track project milestones and timeline on https://github.com/orgs/prestodb/projects/31/
+
+We have the below high-level timeline targets -
```

Comment:
Are there any timeline expectations for jdk 11 upgrade for presto-on-spark? we expect to be ready for this some time in q3.

Reply:
I don't think we have any requirements for a JDK 11 upgrade on the presto-on-spark components on our side at the moment. However, I think we should pull up the minimum version as soon as those deployments are ready to do so. We should also make sure that we socialize that plan to the wider community if that is the case

Reply:
oh wow, i thought this was a requirement.  This is very good to know.

Reply:
Just to clarify: We prototyped a change (https://github.com/prestodb/presto/pull/24395) which would split `presto-main` into two modules. core bits which PoS relies on, and then another module which is responsible for all the Presto server components that rely on Jetty. If we merge that PR the expectation is the the core bits can stay on Java 8 for a short amount of time while users (Meta, etc) take time to upgrade their Spark deployments. Everything else that isn't core would go to Java 17+ ASAP. If we're able to get that PR merged and upgrade all downstream dependencies to Java 17+ I think we will resolve the most critical CVEs on our end. However, we should strive to pull up the language level for the core and PoS bits sooner rather than later.